pkg/diagnose/check.go

```diff
@@ -0,0 +1,122 @@
+
+	// Always disable SBOM collection in `check` command to avoid BoltDB flock issue
+	// and consuming CPU & Memory for asynchronous scans that would not be shown in `agent check` output.
+	pkgconfig.Datadog.Set("container_image_collection.sbom.enabled", "false")
+
+	hostnameDetected, err := hostname.Get(context.TODO())
+	if err != nil {
+		return []diagnosis.Diagnosis{
+			{
+				Result:      diagnosis.DiagnosisFail,
+				Name:        "Host name detection",
+				Diagnosis:   "Failed to get host name and cannot continue to run checks diagnostics",
+				Remediation: "Please validate host environment",
+				RawError:    err,
+			},
+		}
+	}
+
+	// Initializing the aggregator with a flush interval of 0 (to disable the flush goroutines)
+	opts := aggregator.DefaultAgentDemultiplexerOptions()
+	opts.FlushInterval = 0
+	opts.DontStartForwarders = true
+	opts.UseNoopEventPlatformForwarder = true
+	opts.UseNoopOrchestratorForwarder = true
+	forwarder := forwarder.NewDefaultForwarder(config.Datadog, forwarder.NewOptions(config.Datadog, nil))
+	aggregator.InitAndStartAgentDemultiplexer(forwarder, opts, hostnameDetected)
+
+	common.LoadComponents(context.Background(), pkgconfig.Datadog.GetString("confd_path"))
+	common.AC.LoadAndRun(context.Background())
+
+	// Create the CheckScheduler, but do not attach it to
+	// AutoDiscovery.  NOTE: we do not start common.Coll, either.
+	collector.InitCheckScheduler(common.Coll)
+
+	// Load matching configurations (should we use common.AC.GetAllConfigs())
+	waitCtx, cancelTimeout := context.WithTimeout(context.Background(), time.Duration(5*time.Second))
+	diagnoseConfigs, err := common.WaitForAllConfigsFromAD(waitCtx)
+	cancelTimeout()
+	if err != nil {
+		return []diagnosis.Diagnosis{
+			{
+				Result:      diagnosis.DiagnosisFail,
+				Name:        "Check configuration location",
+				Diagnosis:   "Failed to find or load checks configurations",
+				Remediation: "Please validate Agent configuration",
+				RawError:    err,
+			},
+		}
+	}
+
+	// Is there checks to diagnose
+	if len(diagnoseConfigs) == 0 {
+		return nil
+	}
+
+	var diagnoses []diagnosis.Diagnosis
+	for _, diagnoseConfig := range diagnoseConfigs {
+		checkName := diagnoseConfig.Name
+		instances := collector.GetChecksByNameForConfigs(checkName, diagnoseConfigs)
+		for _, instance := range instances {
+
+			instanceName := string(instance.ID())
+
+			// Run check
+			runErr := instance.Run()
+
+			// Get diagnoses
+			checkDiagnoses, diagnoseErr := instance.GetDiagnoses()
+			if diagnoseErr == nil && len(checkDiagnoses) > 0 {
+				for _, d := range checkDiagnoses {
+					if len(d.Category) == 0 {
+						d.Category = instanceName
+					}
+					diagnoses = append(diagnoses, d)
+				}
+			} else if diagnoseErr != nil {
+				// Check Diagnose method return error
+				diagnoses = append(diagnoses, diagnosis.Diagnosis{
+					Result:    diagnosis.DiagnosisUnexpectedError,
+					Name:      instanceName,
+					Category:  instanceName,
```

Comment:
I was wondering what is the purpose of `Category`, if it is always the same as `Name`?

Reply:
The set of Diagnosis structs (Diagnoses) could come from many places. I have added Category fields to potentially differentiate a very large Diagnosis set returned by a singe Diagnose function. In this particular case (for checks), after discussion with Integration team we have decided that they did not want to provide it (seems to be superficial for at least their current though process) hence in the code above I automatically generate it. It is different from a name, for example when a single Check's Diagnose method may provide a number of Diagnoses and they have to be differentiated by name (but their category will be the same). It is important fact that at static/compile time the meta information about a Diagnosis (name and category or description) is not set as far as registration of Diagnose method is concerned.